### PR TITLE
Show diagnostics in scrollbar

### PR DESCRIPTION
**Pull request by bennetbo**
This PR implements support for displaying diagnostics in the scrollbar, similar to what is already done for search results, symbols, git diff, ...

For example, changing a field name (`text`) without changing the references looks like this in `buffer.rs` (note the red lines in the scrollbar):
image

As you can see, the errors, warnings, ... are displayed in the scroll bar, which helps to identify possible problems with the current file.

Relevant issues: #4866, #6819

I'm not so sure about some things:
- What's the performance impact for calling `diagnostics_in_range` on the whole buffer when painting scrollbars? Are there some benchmarks I can run?
- How does scrollbar auto actually work? My scrollbars are always displayed, even when I set them to auto (regardless of my change). So I'm not sure if my change to `element.rs` in line 2171 is correct.
- I don't think the very thin lines that are currently rendered look very nice. Are there any plans to redesign this? At the moment I've just kept the same style as the scrollbar indicators that are already there.

Release Notes:

- Added diagnostic indicators to the scrollbar

### CODE REVIEW

File: crates/multi_buffer/src/multi_buffer.rs [hunk 1]
```diff
@@ -3052,6 +3052,15 @@ impl MultiBufferSnapshot {
         self.has_conflict
     }
 
+    pub fn has_diagnostics(&self) -> bool {
+        for excerpt in self.excerpts.iter() {
```
Comment (SomeoneToIgnore) [inline crates/multi_buffer/src/multi_buffer.rs:3056]: NIT: could be one-liner: `self.excerpts.iter().any(|e| e.buffer.has_diagnostics())`

File: crates/editor/src/element.rs [hunk 2]
```diff
@@ -1477,6 +1478,66 @@ impl EditorElement {
...
+
+                let mut entries = layout
+                    .position_map
+                    .snapshot
+                    .buffer_snapshot
+                    .diagnostics_in_range::<_, Point>(Point::zero()..max_point, false)
+                    .collect::<Vec<_>>();
+
+                // We want to sort by severity, in order to paint the most severe diagnostics last.
+                entries.sort_by_key(|entry| std::cmp::Reverse(entry.diagnostic.severity));
```
Comment (SomeoneToIgnore) [inline crates/editor/src/element.rs:1497]: NIT: I do not think that it changes things drastically, but may be good to remove one obvious allocation by
* removing that `.collect::<Vec<_>>();` entirely
* using itertools' sorting methods (albeit they will use Vec under the hood, still looks more clean?)
* iterating over the iterator directly in `for entry in entries`
* also the name could be `diagnostics`, not some generic `entry`

Up to you though, it's mostly a styling change.

**Review by SomeoneToIgnore — APPROVED** [review]
Great PR, thank you for bringing this in.
I've left a couple of style nit marks, looks good overall.

Answering your questions:

> What's the performance impact for calling diagnostics_in_range on the whole buffer when painting scrollbars? Are there some benchmarks I can run?

We have no benchmarks, but looking at the code, `diagnostics_in_range` goes over the single buffer (as we enable it for singletons only), and uses its `DiagnosticsSet` to retrieve those by ranges and then sort by those ranges.
https://github.com/zed-industries/zed/blob/854480044feb64b5a5c02802b2897bbc11423fb1/crates/language/src/diagnostic_set.rs#L12-L19 hints that we're around `O(n log n)` for all and that sould be ok for most of the cases.
Painting a few extra rectangles onscreen is also what gpui is supposed to deal with well and fast — for the rest, `diagnostics` setting toggle should be enough to fix the perf issues.

We do something similar for buffers already, by highlighting their lines with diagnostics errors, so we should be fine with this PR.
You're welcome to set up a benchmark if you're not convinced.


> How does scrollbar auto actually work? My scrollbars are always displayed, even when I set them to auto (regardless of my change). So I'm not sure if my change to element.rs in line 2171 is correct.

I think your change makes total sense: I am not sure is `auto` broken or not, but worth a standalone check (and fix, if needed). It could appear due to caret placed on some element that got matched in buffer and highlighted.

> I don't think the very thin lines that are currently rendered look very nice. Are there any plans to redesign this? At the moment I've just kept the same style as the scrollbar indicators that are already there.

No plans to redesign this, but something eventually had to be done there: I find them too thin as well, and the scrollbar itself is too narrow to my taste.
If you want to change the design around it, I'd propose to come up with some mockups and a separate issue for discussing it.

Soo, let's merge the nits and get it merged?

**Comment by bennetbo** [comment]
Thanks for the review, explanation and the great suggestions. I did some cleanup.

**Review by SomeoneToIgnore — APPROVED** [review]

**Pull request merged**